pilot/pkg/model/context.go

```diff
@@ -332,6 +332,14 @@ type Proxy struct {
 	// LastPushContext; the XDS cache depends on knowing the time of the PushContext to determine if a
 	// key is stale or not.
 	LastPushTime time.Time
+
+	// LastPushedResources contains the list of resources that were last time to the proxy,
+	// keyed by the DiscoveryRequest TypeUrl. Currently it is used only for CDS
+	LastPushedResources map[string][]*discovery.Resource
```

Comment:
This won't scale, the cost to store all of these is crazy high. CDS response for a single proxy can easily be 10mb

Reply:
But I also don't think we need it  - we can easily regenerate the 'removed' clusters. See https://github.com/howardjohn/istio/tree/gateway/cluster-scope where I tried this (but not the ResourceVersionWaitingForAck part, which is why it didn't work).

Reply:
The PR adds a test as well btw

Reply:
Ok, thanks. I just wanted to be sure that you didn't solve that case.
Did you consider persisting state by writing previous clusters to the status of VirtualService?

Reply:
I did briefly but I don't think there is a feasible way to do it. 1 VS may apply to 100 GWs all with a different set of clusters, and the cost to maintain that field is probably extremely high. IMO envoy should just support this natively; pushing hacks further up the stack generally causes more pain.

Reply:
Ok, thanks.

> IMO envoy should just support this natively; pushing hacks further up the stack generally causes more pain.

Yes, I totally agree. I just look for a temporary fix, because implementing this in Envoy will be much more time consuming.